### PR TITLE
Check rendererCanHaveTrimmedMargin before calling box->hasTrimmedMargin

### PR DESCRIPTION
**Pull request by sammygill**
#### beefa0360c1e5611074acc823e60878b9be936db
<pre>
Check rendererCanHaveTrimmedMargin before calling box-&gt;hasTrimmedMargin
<a href="https://bugs.webkit.org/show_bug.cgi?id=254859">https://bugs.webkit.org/show_bug.cgi?id=254859</a>
rdar://problem/107502795

Reviewed by Alan Baradlay.

In this example we will trigger ASSERT(!needsLayout()) inside of
RenderBox::hasTrimmedMargin. This is because computing the margin on
the item is not layout dependent (isLayoutDependent returns false)
and we end up calling into box-&gt;hasTrimmedMargin before layout has
occurred.

&lt;style&gt;
.flexbox {
    display: flex;
}
.flex-item {
    width: 50px;
    height: 50px;
    margin-top: 10px;
    background-color: green;
}
&lt;/style&gt;
&lt;body&gt;
    &lt;div class=&quot;flexbox&quot;&gt;
        &lt;div id=&quot;flex-item&quot; class=&quot;flex-item&quot;&gt;&lt;/div&gt;
    &lt;/div&gt;
&lt;script&gt;
    window.getComputedStyle(document.getElementById(&quot;flex-item&quot;)).marginTop;
&lt;/script&gt;
&lt;/body&gt;

Before calling hasTrimmedMargin we should check for the same conditions
that would make isLayoutDependent return true. To accomplish this, I
renamed rendererContainingBlockHasMarginTrim to rendererCanHaveTrimmedMargin
and removed the asserts that were in there. These asserts were
unnecessary because they were the same ones being checked inside of
hasTrimmedMargin. Now we can use this helper function inside of
isLayoutDependent and to guard the call to hasTrimmedMargin. Another
benefit of this is that as we iterate on the computed values for
trimmed margins in different layout contexts we can just change this
one function rather than sprinkling the same checks around each of
the margin types.

* LayoutTests/css3/flexbox/flex-item-margin-top-no-crash-expected.html: Added.
* LayoutTests/css3/flexbox/flex-item-margin-top-no-crash.html: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::rendererCanHaveTrimmedMargin):
(WebCore::isLayoutDependent):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
(WebCore::rendererContainingBlockHasMarginTrim): Deleted.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::hasTrimmedMargin const):

Canonical link: <a href="https://commits.webkit.org/262679@main">https://commits.webkit.org/262679@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/89083d42cb1490260918245f9cf827245fea99c8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2215 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2250 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2319 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3141 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2255 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2188 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/2365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2297 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2002 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2235 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1980 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1987 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2998 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/1958 "Build was cancelled. Recent messages:OS: Monterey (12.6), Xcode: 14.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests (cancelled)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1966 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1859 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2047 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2013 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3162 "268 api tests failed or timed out") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2033 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1822 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1960 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1975 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/552 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2151 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 1]
```diff
@@ -2342,10 +2333,10 @@ static bool isLayoutDependent(CSSPropertyID propertyID, const RenderStyle* style
             return false;
         return !(style && style->marginTop().isFixed() && style->marginRight().isFixed()
             && style->marginBottom().isFixed() && style->marginLeft().isFixed())
-            || (isFlexItem(renderer) && rendererContainingBlockHasMarginTrim(downcast<RenderBox>(*renderer), { }));
+            || canCheckForTrimmedMargin(downcast<RenderBox>(*renderer), { });
```
Comment (alanbaradlay) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: canCheckForTrimmedMargin is a somewhat confusing name in this context.
```
bool isLayoutDependent()
{
    if (canCheckForFoo())
        return hasFoo();
...
```
is what I would expect as opposed to
```
bool isLayoutDependent()
{
    return canCheckForFoo();
}
```

Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: I went ahead and moved this function outside of `ComputedStyleExtractor` and  into `RenderBox`. I also moved the responsibility of checking this away from the caller and into `hasTrimmedMargin` itself so that we can avoid pairing these function calls together at each call site. Let me know what you think

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 2]
```diff
@@ -2342,10 +2314,10 @@ static bool isLayoutDependent(CSSPropertyID propertyID, const RenderStyle* style
             return false;
         return !(style && style->marginTop().isFixed() && style->marginRight().isFixed()
             && style->marginBottom().isFixed() && style->marginLeft().isFixed())
-            || (isFlexItem(renderer) && rendererContainingBlockHasMarginTrim(downcast<RenderBox>(*renderer), { }));
+            || downcast<RenderBox>(*renderer).canCheckForTrimmedMargin({ });
```
Comment (alanbaradlay) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: Is this a leftover?

File: Source/WebCore/rendering/RenderBox.cpp [hunk 3]
```diff
@@ -1434,41 +1434,50 @@ void RenderBox::clearTrimmedMarginsMarkings()
...
-        return false;
-    return marginTrimType ? (rareData().trimmedMargins() & static_cast<unsigned>(*marginTrimType)) : rareData().trimmedMargins();
+
+    if (isFlexItem() && (!marginTrimType || marginTrimType == MarginTrimType::BlockStart))
+        return marginTrimType ? containingBlock->style().marginTrim().contains(*marginTrimType) : !containingBlock->style().marginTrim().isEmpty();
+    return false;
 }
 
-bool RenderBox::hasTrimmedMargin(PhysicalDirection physicalDirection) const
+bool RenderBox::hasTrimmedMargin(std::variant<std::monostate, PhysicalDirection, MarginTrimType> direction) const
```
Comment (alanbaradlay) [inline Source/WebCore/rendering/RenderBox.cpp]: This is getting a bit too difficult to parse. Are you sure we need variant to solve this problem?

Comment (sammygill) [reply Source/WebCore/rendering/RenderBox.cpp]: Nope. It would have been nice to get rid of the function overloading imo but not if it worsens the readability of the code. Went back to my original approach and am now just adding the extra check before calling `box->hasTrimmedMargin`

It would be nice to do some clean up of the code, but that might be better in a separate patch when all of the similar logic for each side has been implemented (at least for flex layout)

Also I think this would be a good opportunity to do the function renaming like you mentioned in https://github.com/WebKit/WebKit/pull/11741

The new name is a little verbose, but hopefully it gets the point across and is accurate in its name

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 4]
```diff
@@ -2291,8 +2291,13 @@ static bool isFlexItem(const RenderObject* renderer)
     return false;
 }
 
-static bool rendererContainingBlockHasMarginTrim(const RenderBox& renderer, std::optional<MarginTrimType> marginTrimType)
+static bool rendererCanHaveMarginTrimRareDataSetWithinFormattingContext(const RenderBox& renderer, std::optional<MarginTrimType> marginTrimType)
 {
+    // A renderer will have a specific margin marked as trimmed by setting its rare data bit if:
+    // 1.) The layout system the box is in has this logic (setting the rare data bit for this 
+    // specific margin) implemented
+    // 2.) The block container/flexbox/grid has this margin specified in its margin-trim sty;e
```
Comment (stwrt) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: Should be `style`.

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 5]
```diff
@@ -2306,7 +2311,9 @@ static bool rendererContainingBlockHasMarginTrim(const RenderBox& renderer, std:
 
     if (containingBlock->isFlexibleBox()) {
         ASSERT(!marginTrimType || *marginTrimType == MarginTrimType::BlockStart);
```
Comment (stwrt) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: I am not sure what this ASSERT is buying us exactly. We only call this function in two instances with `{}` or `BlockStart`. This will always be true.

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 6]
```diff
@@ -3279,7 +3286,9 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
             return CSSPrimitiveValue::create(CSSValueAuto);
         return CSSPrimitiveValue::createCustomIdent(style.specifiedLocale());
     case CSSPropertyMarginTop: {
-        if (auto* box = dynamicDowncast<RenderBox>(renderer); box && box->isFlexItem() && box->hasTrimmedMargin(PhysicalDirection::Top))
+        if (auto* box = dynamicDowncast<RenderBox>(renderer); box && box->isFlexItem()
```
Comment (stwrt) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: can you move the first part `auto *box =` out of the if statement.

**Review by stwrt — COMMENTED** [review]
A few comments from an initial pass through.

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 6 above)
Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: Done

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 4 above)
Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: Fixed

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 5 above)
Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: Yeah now that you bring it up this may be a bit overkill. I'm also thinking that all of these ASSERTs within this function are a bit overkill. The purpose is to return true when `box->hasTrimmedMargin()` can be called only on supported content so that it doesn't assert. Having this function ASSERT instead of just simply returning false might not be the best choice here. Do you both agree and think that they can be removed from this function? @stwrt @alanbaradlay

Comment (stwrt) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: My personal opinion is to remove it.

Comment (stwrt) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]:
```suggestion
```

🙂

File: Source/WebCore/rendering/RenderBox.cpp [hunk 7]
```diff
@@ -1455,14 +1463,6 @@ bool RenderBox::hasTrimmedMargin(PhysicalDirection physicalDirection) const
 {
     if (!isInFlow())
         return false;
-#if ASSERT_ENABLED
```
Comment (stwrt) [inline Source/WebCore/rendering/RenderBox.cpp]: This seems like it could just be a separate commit for code cleanup...

Comment (sammygill) [reply Source/WebCore/rendering/RenderBox.cpp]: Sounds good to me. Added the code block back

File: Source/WebCore/rendering/RenderBox.cpp [hunk 8]
```diff
@@ -1439,12 +1439,19 @@ bool RenderBox::hasTrimmedMargin(std::optional<MarginTrimType> marginTrimType) c
...
 #if ASSERT_ENABLED
+    // We should assert here if this function is called with a layout system and
+    // MarginTrimType combination that is not supported yet (i.e. the layout system
+    // does not set the margin trim rare data bit for that margin)
+
     // containingBlock->isBlockContainer() can return true even if the item is in a RenderFlexibleBox
     // (e.g. buttons) so we should explicitly check that the item is not a flex item to catch block containers here
-    if (auto* containingBlock = this->containingBlock(); containingBlock && !containingBlock->isFlexibleBox() &&  (containingBlock->isBlockContainer() || containingBlock->isRenderGrid())) {
+    auto* containingBlock = this->containingBlock(); 
+    if (containingBlock && !containingBlock->isFlexibleBox() &&  (containingBlock->isBlockContainer() || containingBlock->isRenderGrid())) {
```
Comment (stwrt) [inline Source/WebCore/rendering/RenderBox.cpp]: extra space

Comment (sammygill) [reply Source/WebCore/rendering/RenderBox.cpp]: Removed

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 9]
```diff
@@ -2291,24 +2291,27 @@ static bool isFlexItem(const RenderObject* renderer)
     return false;
 }
 
-static bool rendererContainingBlockHasMarginTrim(const RenderBox& renderer, std::optional<MarginTrimType> marginTrimType)
+static bool rendererCanHaveMarginTrimRareDataSetWithinFormattingContext(const RenderBox& renderer, std::optional<MarginTrimType> marginTrimType)
```
Comment (alanbaradlay) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: this is mouthful. RareData is an internal concept, should not appear in this function name.

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 10]
```diff
@@ -2291,24 +2291,27 @@ static bool isFlexItem(const RenderObject* renderer)
...
 {
+    // A renderer will have a specific margin marked as trimmed by setting its rare data bit if:
+    // 1.) The layout system the box is in has this logic (setting the rare data bit for this 
+    // specific margin) implemented
+    // 2.) The block container/flexbox/grid has this margin specified in its margin-trim style
+    // If marginTrimType is empty we will check if any of the supported margins are in the style
     auto* containingBlock = renderer.containingBlock();
     if (!containingBlock || containingBlock->isRenderView())
         return false;
```
Comment (alanbaradlay) [inline Source/WebCore/css/ComputedStyleExtractor.cpp:2303]: Any reason for checking containing block even before establishing that we are dealing with BFC here?

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 11]
```diff
@@ -3279,7 +3282,9 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
             return CSSPrimitiveValue::create(CSSValueAuto);
         return CSSPrimitiveValue::createCustomIdent(style.specifiedLocale());
     case CSSPropertyMarginTop: {
-        if (auto* box = dynamicDowncast<RenderBox>(renderer); box && box->isFlexItem() && box->hasTrimmedMargin(PhysicalDirection::Top))
+        auto* box = dynamicDowncast<RenderBox>(renderer);
+        if (box && rendererCanHaveMarginTrimRareDataSetWithinFormattingContext(*box, MarginTrimType::BlockStart) 
+            && box->hasTrimmedMargin(PhysicalDirection::Top))
```
Comment (alanbaradlay) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: any particular reason for changing box's scope here? Are you planning on using it outside?

File: Source/WebCore/rendering/RenderBox.cpp [hunk 12]
```diff
@@ -1439,12 +1439,19 @@ bool RenderBox::hasTrimmedMargin(std::optional<MarginTrimType> marginTrimType) c
...
 #if ASSERT_ENABLED
+    // We should assert here if this function is called with a layout system and
+    // MarginTrimType combination that is not supported yet (i.e. the layout system
+    // does not set the margin trim rare data bit for that margin)
+
     // containingBlock->isBlockContainer() can return true even if the item is in a RenderFlexibleBox
     // (e.g. buttons) so we should explicitly check that the item is not a flex item to catch block containers here
-    if (auto* containingBlock = this->containingBlock(); containingBlock && !containingBlock->isFlexibleBox() &&  (containingBlock->isBlockContainer() || containingBlock->isRenderGrid())) {
+    auto* containingBlock = this->containingBlock(); 
+    if (containingBlock && !containingBlock->isFlexibleBox() && (containingBlock->isBlockContainer() || containingBlock->isRenderGrid())) {
```
Comment (alanbaradlay) [inline Source/WebCore/rendering/RenderBox.cpp:1464]: I should have noticed this incorrect use of "containingBlock" in the original PR. While it's ok to land as is, at some point (very low priority) this needs a cleanup/stop using containing block outside of BFC.

**Review by alanbaradlay — APPROVED** [review]

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 5 above)
Comment (alanbaradlay) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: > Yeah now that you bring it up this may be a bit overkill. I'm also thinking that all of these ASSERTs within this function are a bit overkill. The purpose is to return true when `box->hasTrimmedMargin()` can be called only on supported content so that it doesn't assert. Having this function ASSERT instead of just simply returning false might not be the best choice here. Do you both agree and think that they can be removed from this function? @stwrt @alanbaradlay

ASSERTs saved me from introducing embarrassing regressions many times in the past. Because of that I am all in favor of defensive programming and I do tend to overuse them. However I also understand that not everyone finds them useful so while I don't necessarily agree, I don't mind not having them either.

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 10 above)
Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp:2303]: I believe this was related to some test cases there were failing with a null containing block (so I added it after seeing the test results). The logic here should probably be rearranged and should attempt to first figure out the formatting context before attempting a check like this. I can do this in the same patch when I clean up the incorrect uses of `containingBlock` (unless you would rather have it fixed here)

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 9 above)
Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: Changed to `rendererCanHaveTrimmedMargin`

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 11 above)
Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: We were having a conversation about readability with this particular piece of code. I'm ok with adding it back though
